### ISSUE TITLE
"Iterable bodies must always iterate in bytestrings"

### ISSUE THREAD

**Issue by thoschi** [KnugiHK/synapse-admin-api-python#1]
I just started testing your wrapper.

So far I can connect to our matrix server, get information (e.g. version, user list).
As soon as I try to send a body to the server (i.e. to create a user, send an announcement, ...), I get an error:

"Iterable bodies must always iterate in bytestrings"

Maybe I am just doing something wrong - is there any documentation or could you provide an example of user creation?

Thanks and best regards
Thomas

**Comment by KnugiHK** [comment]
Yes, I will do some documentation later.

JSON keys in the body should be treated as arguments, for example, user creation:
```python
from synapse_admin import User
user = User()
print(user.creates("test001", password="password", displayname="Test")) # True should be returned.
print(user.lists()) # List out all users
```
Did you pass a dict or json into the function? Just like:
```python
body = {"password": "password", "displayname": "Test"}
user.creates("test001", body)
```
I am still thinking whether the function should accept a dict body or treat the body as arguments, comments are welcome.

Btw, I fixed a minor bug in the function of creating user, which do not affect the succession of user creation. You may want to git clone it and install it again, release will be created soon.

**Comment by KnugiHK** [comment]
It will be appreciated if you could provide me the code that can re-produce the mentioned exception for me to do some more investigation. Thanks!

**Comment by thoschi** [comment]
Thanks for your reply!

First it seems that I basically understood what to do. Anyway your hints allowed me to try some more. Still the results are the same.

So far I tried both ways you mentioned - as kwargs (I realized the json.dumps inside) and as a pre-jsonified body.
I tried the pip version as well as the latest github-version.
My Python version is 3.7.10
I will add the full output at the end.

Something else I observed: shouldn't it be possible to create e.g. a User-instance passing the variables for server/port/token? If I try, it anyway asks me for the data I submitted.

The full error message is:
Traceback (most recent call last):
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/hyper/http11/connection.py", line 333, in _send_chunked
    self._sock.send(chunk)
  File "/usr/lib/python3.7/ssl.py", line 1003, in send
    return self._sslobj.write(data)
TypeError: a bytes-like object is required, not 'str'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 5, in <module>
    print(user.creates("test11", password="xxxxxxxxx", displayname="Test"))
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/matrix_synapse_admin-0.1.1-py3.7.egg/synapse_admin/user.py", line 99, in creates
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/hyper/common/connection.py", line 103, in request
    method=method, url=url, body=body, headers=headers
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/hyper/http11/connection.py", line 187, in request
    self._send_body(body, body_type)
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/hyper/http11/connection.py", line 317, in _send_body
    return self._send_chunked(body)
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/hyper/http11/connection.py", line 337, in _send_chunked
    "Iterable bodies must always iterate in bytestrings"
ValueError: Iterable bodies must always iterate in bytestrings

Best regards
Thomas

**Comment by KnugiHK** [comment]
Alright, I can reproduce the error you mentioned. It seems that it is related to the HTTP version. When I do it through my reverse proxy, which enables HTTP2, I created a user successfully. However, when I do it without reverse proxy (no http2), the exception is raised. Probably it is because the HTTP2 use bytes stream and HTTP1 use bytestring to communicate? I will do further investigation on it.

And for the User-instance passing, the variables and the program still ask for the information. I found that it is a bug. I will do some investigation and fix it later.

**Comment by KnugiHK** [comment]
I found that in the hyper library, which acts as an HTTP client, it will convert string to bytestring in HTTP2 but not in HTTP1, therefore, when sending a request in HTTP1, the body must be bytestring before passing it to hyper. And... I just realized Hyper is no longer maintained. I will move it to another HTTP client library soon.

The bug of User-instance passing variables will be fixed after the migration complete.

**Comment by KnugiHK** [comment]
The HTTP client is migrated from Hyper to HTTPX, you can try to git clone or git pull it again to install the latest version, the User.creates should now work on HTTP/1. However, testing is needed for almost the entire library since they are not being tested after migration.

**Comment by thoschi** [comment]
First quick test - user creation now works!

But I encountered a "new" error when calling `user.lists()`:

```
File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/matrix_synapse_admin-0.1.1-py3.7.egg/synapse_admin/user.py", line 88, in lists
synapse_admin.base.SynapseException: SynapseException: [M_UNKNOWN] Boolean query parameter b'deactivated' must be one of ['true', 'false']

```

**Comment by KnugiHK** [comment]
Oh, a little bug, fixed.

**Comment by thoschi** [comment]
It seems to work now - thanks a lot! Few more bugs/ideas:

When calling `room.lists()`:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/schulzeug-matrix/env/lib/python3.7/site-packages/matrix_synapse_admin-0.1.1-py3.7.egg/synapse_admin/room.py", line 96, in lists
TypeError: request() got an unexpected keyword argument 'headers'
```

Another question: is the creation of rooms just inherited from base/Admin class - did not find anything in room.py. I did not yet manage to create a room, but just started trying.
Btw. if I call room.create() it asks me for the server data each time (but this might be due to misusage).

I realized that the settings, once entered, are saved to a file. Do you think, it is possible to pass a switch to disable this? I can think of situations, where one dont want to have something written to a home directory. I would even try a pull request but I understood you first want to check the problem mentioned in top post.

Best regards
Thomas

**Comment by KnugiHK** [comment]
> TypeError: request() got an unexpected keyword argument 'headers'

Yeah, probably the bug comes from the migration of the HTTP client. Fixed earlier, try to git pull it.

Yes, in the current design, the program allows the developers to pass the connection information to the class in which the information will only be used once. Meaning that if you pass the server_addr and access_token into the program, no config will involve. However, some bugs make the config part involved.

**Comment by KnugiHK** [comment]
The connection information passing is now fixed. You may try to see if it does what you want.
```python
from synapse_admin import User
user = User("example.com", access_token="<JWT>")
print(user.lists())
```
This will not ask for further information now and they are not stored in config file.

**Comment by thoschi** [comment]
Hi - no more errors so far. Great! Thanks a lot.
Also I like the logic you described earlier - best of both worlds, depending how one uses it.

So only remaining question so far is room creation.

If I create an instance of the Room class and call `room.create()` it asks (again) for server name and so on. But actually I am not sure what parameters to choose. "<something>" does not work. But kwargs ("canonical_name=<something>") raises `TypeError: create() got an unexpected keyword argument 'canonical_alias'`.

But I do not even know if it is meant to work that way.

Thanks so far for your works! We are working on a small middleware between our school server and different services, so that all users (from different sources) are registered to these services and enrolled in courses/groups/...). To have an internal matrix server with all courses ready at the beginning of the school year would be another nice accomplishment. And using your wrapper is a very welcome short cut on the way :-)

**Comment by KnugiHK** [comment]
Glad to know that this wrapper helps your work! I also planned to develop the CLI and Telegram bot version base on this wrapper, check them out later if you want.

For the room creation, this wrapper is currently based on [the documentation of Synapse Admin API](https://github.com/matrix-org/synapse/tree/develop/docs/admin_api), therefore, if the API endpoint is not included in the documentation, it is not implemented in this wrapper yet. If you know there are some undocumented endpoints, More issues are welcome! I will also try to figure out all API endpoints.

Btw, since class Room inherits from class Admin, therefore, you are actually calling the create() function in Admin, which is used for creating the configuration. I think I should change the name of the configuration creation function to avoid confusion.

**Comment by thoschi** [comment]
Yes - as I wrote I suggested, that the create method I called was inherited from Admin() and surely not what I wanted to do. I just wondered, why I am asked for the server information. Anyway it is of course not meant to be used that way.

As for room creation - it was my mistake. I read about "create_room" but this is part of the client-api. And for that there is the client-api available (https://github.com/matrix-org/matrix-python-sdk).

So for now it seems I can do all I want to do. I have another project first (a document camera application for digital whiteboards), after that I am looking forward to write the matrix-extension as planned.

Thank you for your time and work! I will for sure check out the progress from time to time.

Best regards from Germany
Thomas

**Comment by KnugiHK** [comment]
I renamed some functions in Admin() to avoid confusion. Room.create() is no longer exist.

Regarding room creation, it may be a good idea to include some client APIs that are not implemented in the Admin API in this wrapper. I will take a look into it. Thanks for your contribution! And I hope that your project will be successful!

Btw, I have added an option for the user to choose whether save the connection information or not after entering the information.
